src/harddisk/pata.rs

```diff
@@ -1,37 +1,229 @@
+use core::sync::atomic::AtomicBool;
+
+use x86_64::instructions::port::{Port, PortReadOnly, PortWriteOnly};
+
 use crate::svec::SVec;
 
+// Assuming "typical" ports, LBA48
+const IO_BASE_PORT: u16 = 0x1F0;
+static mut DATA_REG: Port<u16> = Port::new(IO_BASE_PORT);
+static mut ERROR_REG: PortReadOnly<u16> = PortReadOnly::new(IO_BASE_PORT + 1);
+static mut FEATURES_REG: PortWriteOnly<u16> = PortWriteOnly::new(IO_BASE_PORT + 1);
+static mut SECTOR_COUNT_REG: Port<u8> = Port::new(IO_BASE_PORT + 2); //Actually u16, but low and high needs to be sent separately.
+static mut LBA_LOW_REG: Port<u8> = Port::new(IO_BASE_PORT + 3); //same
+static mut LBA_MID_REG: Port<u8> = Port::new(IO_BASE_PORT + 4); // with these 2
+static mut LBA_HIGH_REG: Port<u8> = Port::new(IO_BASE_PORT + 5);
+static mut DRIVE_HEAD_REG: Port<u8> = Port::new(IO_BASE_PORT + 6);
+static mut STATUS_REG: PortReadOnly<u8> = PortReadOnly::new(IO_BASE_PORT + 7);
+static mut COMMAND_REG: Port<u8> = Port::new(IO_BASE_PORT + 7);
+const CONTROL_BASE_PORT: u16 = 0x3F6;
+static mut ALT_STATUS_REG: PortReadOnly<u8> = PortReadOnly::new(CONTROL_BASE_PORT);
+static mut DEVICE_CONTROL_REG: PortWriteOnly<u8> = PortWriteOnly::new(CONTROL_BASE_PORT + 0);
+static mut DRIVE_ADRESS_REG: PortReadOnly<u8> = PortReadOnly::new(CONTROL_BASE_PORT + 0);
+
+/// Is the driver/disk busy?
+/// TODO: make this per disk
+static BUSY: AtomicBool = AtomicBool::new(false);
+
+static mut IDENTIFY_RESULT: [u16; 256] = [0; 256];
 
 pub unsafe fn initialize() {
-    todo!()
+    let status = STATUS_REG.read();
+    if status == 0xFF {
+        panic!("Floating bus");
+    } else {
+        DRIVE_HEAD_REG.write(0xA0);
+        send_lba_and_sector_count(0, 0);
+        COMMAND_REG.write(0xEC); //IDENTIFY
+        let status = STATUS_REG.read();
+        if status == 0 {
+            panic!("Drive does not exist");
+        }
+        while STATUS_REG.read() & 0x80 != 0 {}
+        if LBA_MID_REG.read() != 0 || LBA_HIGH_REG.read() != 0 {
+            panic!("Non-ATA drive");
+        }
+        loop {
+            let status = STATUS_REG.read();
+            if status & 8 == 0 {
```

Comment:
Shouldn't this be `0x80`? according to oswiki

Reply:
Actually, no, i read wrong :P

Reply:
I did read wrong, but there is still something wrong.
It should probs be `== 8`